### PR TITLE
 Adding copy method to ChainReader

### LINKED ISSUE #1784: Add copy method to ChainReader
The `coordinates.ChainReader` class needs to implement a copy method.  This should work identically to the base Reader's copy method, but also take into account any peculiarities of the ChainReader

### PR DESCRIPTION
**Pull request by navyakhare**
Fixes #1784 

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/coordinates/chain.py [hunk 1]
```diff
@@ -47,6 +47,7 @@
 import os.path
 import itertools
 import bisect
+import copy
```
Comment (richardjgowers) [inline package/MDAnalysis/coordinates/chain.py]: You don't need to import `copy` here, we're not using it.  `copy` would be needed if you called `copy.something()` not for calling `thing.copy()`

File: package/MDAnalysis/coordinates/chain.py [hunk 2]
```diff
@@ -211,7 +212,19 @@ def convert_pos_to_native(self, x):
         return self.active_reader.convert_pos_to_native(x)
 
     def copy(self):
-        raise NotImplementedError("Copy not implemented for ChainReader")
+        #raise NotImplementedError("Copy not implemented for ChainReader")
+        new = self.__class__(self.filenames)
```
Comment (richardjgowers) [inline package/MDAnalysis/coordinates/chain.py]: If you're passing in the filenames here, you probably don't have to then copy the Readers.  I'd say you can do one or the other, so maybe see if you can get away with just passing filenames.

You might have to `copy.copy(self.filenames)` else they both share a reference to the same Python list

File: package/MDAnalysis/coordinates/chain.py [hunk 3]
```diff
@@ -211,7 +212,19 @@ def convert_pos_to_native(self, x):
...
+        new.readers=[reader.copy() for reader in self.readers]
+        new.n_atoms = self._get_same('n_atoms').copy()
+        n_frames = self._get('n_frames').copy()
+        new.__start_frames = np.cumsum([0] + n_frames)
+        new.n_frames = np.sum(n_frames)
+        new.dts = np.array(self._get('dt').copy())
+        new.total_times = new.dts * n_frames
+        new.rewind()
```
Comment (richardjgowers) [inline package/MDAnalysis/coordinates/chain.py]: Don't rewind, the new reader instead needs to be at the same point as `self`

File: package/MDAnalysis/coordinates/chain.py [hunk 4]
```diff
@@ -211,7 +212,19 @@ def convert_pos_to_native(self, x):
         return self.active_reader.convert_pos_to_native(x)
 
     def copy(self):
-        raise NotImplementedError("Copy not implemented for ChainReader")
+        #raise NotImplementedError("Copy not implemented for ChainReader")
+        new = self.__class__(self.filenames)
+        new.readers=[reader.copy() for reader in self.readers]
+        new.n_atoms = self._get_same('n_atoms').copy()
```
Comment (richardjgowers) [inline package/MDAnalysis/coordinates/chain.py]: I don't think setting `n_atoms` `n_frames` etc is necessary.  The new Reader can work these out the same way as this Reader does

**Comment by navyakhare** [comment]
I have started working on the correction and the tests. I have mid term exams for the next four days and hence won't be very active. I'll immediately get back to you with the tests after it. Apologies for the delay :( .

**Comment by richardjgowers** [comment]
@navyakhare it's a little confusing, but there's a `pytest.fixture` here: https://github.com/MDAnalysis/mdanalysis/blob/develop/testsuite/MDAnalysisTests/coordinates/test_copying.py#L61

which is like a for loop over all possible Reader objects.  Then every test in the same file which uses the `original_and_copy` fixture loops over all Readers in that list.

If you add a new line with `    ('CHAIN', [GRO, GRO, GRO], dict()),` to that fixture, it will then run ChainReader through all the same tests as every other format.  Doing that makes most tests pass, but there's 3 that don't.  I think if you see which tests fail and compare the copy method you've implemented to the base one you'll see what changes are necessary to get all tests passing.

**Comment by navyakhare** [comment]
@richardjgowers  thanks for the explanation. All test cases are passing now. While writing test cases in `test_chainreader.py` I was taking `BaseReaderTest` as reference. But it looks like BaseReader is not testing its copy method via `test.copying.py`. Why is that so? I can understand that all the tests cant be generalized in test_copying but how are we differentiating between these tests?

File: testsuite/MDAnalysisTests/coordinates/test_chainreader.py [hunk 5]
```diff
@@ -116,6 +116,55 @@ def test_write_dcd(self, universe, tmpdir):
                 err_msg="Coordinates disagree at frame {0:d}".format(
                     ts_orig.frame))
 
+    def test_copy_basic(self, universe):
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/coordinates/test_chainreader.py]: I think these tests are redundant now you've got the other tests working

File: testsuite/MDAnalysisTests/coordinates/test_copying.py [hunk 6]
```diff
@@ -187,13 +188,11 @@ def test_positions_share_memory(original_and_copy):
     with pytest.raises(AssertionError):
         assert_equal(original.ts.positions, copy.ts.positions)
 
-def test_chainreader_NIE():
-    # ChainReader not implemented, check error message is sane
+def test_chainreader():
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/coordinates/test_copying.py]: You can just remove this test entirely

**Comment by richardjgowers** [comment]
@navyakhare we can't use BaseReader as (I think?) every Reader isn't using that yet, so when I wrote the copying tests I instead just created a whole new file for them.

**Comment by navyakhare** [comment]
@richardjgowers  yes my name is already there. Should I add this change to CHANGELOG?

**Comment by richardjgowers** [comment]
@navyakhare thanks for the fix!  Sometimes it's easiest to work backwards from failing tests!

**Comment by navyakhare** [comment]
@richardjgowers  Yes indeed :D !

**Pull request merged**